src/ImageSharp/Dithering/ErrorDiffusion/ErrorDiffuser.cs

```diff
@@ -0,0 +1,107 @@
+﻿// <copyright file="ErrorDiffuser.cs" company="James Jackson-South">
+// Copyright (c) James Jackson-South and contributors.
+// Licensed under the Apache License, Version 2.0.
+// </copyright>
+
+namespace ImageSharp.Dithering
+{
+    using System;
+    using System.Numerics;
+    using System.Runtime.CompilerServices;
+
+    /// <summary>
+    /// The base class for performing effor diffusion based dithering.
+    /// </summary>
+    public abstract class ErrorDiffuser : IErrorDiffuser
+    {
+        /// <summary>
+        /// The vector to perform division.
+        /// </summary>
+        private readonly Vector4 divisorVector;
+
+        /// <summary>
+        /// The matrix width
+        /// </summary>
+        private readonly byte matrixHeight;
+
+        /// <summary>
+        /// The matrix height
+        /// </summary>
+        private readonly byte matrixWidth;
+
+        /// <summary>
+        /// The offset at which to start the dithering operation.
+        /// </summary>
+        private readonly int startingOffset;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ErrorDiffuser"/> class.
+        /// </summary>
+        /// <param name="matrix">The dithering matrix.</param>
+        /// <param name="divisor">The divisor.</param>
+        protected ErrorDiffuser(byte[,] matrix, byte divisor)
+        {
+            Guard.NotNull(matrix, nameof(matrix));
+            Guard.MustBeGreaterThan(divisor, 0, nameof(divisor));
+
+            this.Matrix = matrix;
+            this.matrixWidth = (byte)(matrix.GetUpperBound(1) + 1);
+            this.matrixHeight = (byte)(matrix.GetUpperBound(0) + 1);
+            this.divisorVector = new Vector4(divisor);
+
+            this.startingOffset = 0;
+            for (int i = 0; i < this.matrixWidth; i++)
+            {
+                if (matrix[0, i] != 0)
+                {
+                    this.startingOffset = (byte)(i - 1);
+                    break;
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public byte[,] Matrix { get; }
```

Comment:
`float[][]` is still crazy slow. We need at least [something like this](https://gist.github.com/antonfirsov/6c11a2f0a1048e8bcca455c072d14810). 

Even better to have simple on-stack value types for these small matrices. (Need to invent a trick for this).